### PR TITLE
added new distributions to sampling operators. added operators for sa…

### PR DESCRIPTION
**Pull request by asmushetzel**
…mpling multiple distributions that are parametrized by input arguments(tensors)

This pull request depends on PR 228 for mshadow! 
I'm insecure how we proceed as this PR will fail as long as PR 228 has not yet been pulled. 

This branch adds methods to sample from more often used distributions: Exponential/Gamma/Poisson/Negative Binomial. All these sampling methods are currently only available for CPU. They are available for ndarray and symbolic APIs. Moreover a set of "multi-sample" operators is available where distribution parameters are tensor inputs. With this, MxNet can be used to estimate parameters of a probability distribution by a neural network and then sample from this distribution and feed the sampling then into other parts of a neural network. That is a critical functionality used for a lot of applications and similar to what TensorFlow offers. 
Tests have been added to test_random.py for all these changes.
In the future, we may think about extending and unifying the random samplers further. Goal should be to have additional operators in there for example for likelihood. And provide GPU versions for all distribution types.

### CODE REVIEW

File: src/operator/tensor/sample_op.h [hunk 1]
```diff
@@ -35,13 +36,8 @@ struct SampleUniformParam : public dmlc::Parameter<SampleUniformParam> {
     .describe("Context of output, in format [cpu|gpu|cpu_pinned](n)."
               "Only used for imperative calls.");
     DMLC_DECLARE_FIELD(dtype)
-    .add_enum("None", -1)
```
Comment (piiswrong) [inline src/operator/tensor/sample_op.h]: ?

Comment (asmushetzel) [inline src/operator/tensor/sample_op.h]: Apparently this was introduced by PR 5019 end of Feb and my changes had been build on top of a version right before that PR. Will merge all of this correctly.

Comment (piiswrong) [inline src/operator/tensor/sample_op.h]: try git fetch upstream && git rebase upstream/master

**Comment by asmushetzel** [comment]
There is a slight  naming issue for the simple sampling operators in sample_op.cc:
The prior existing ones had been named "uniform" and "normal". Going forward with the same naming scheme would add "gamma", "poisson" etc. But "gamma" is already taken as an operator evaluating the gamma function. So decided to go with the naming scheme "random_uniform", "random_normal", random_gamma" which is a bit more expressive anyway. To preserve backward compatibility, there is an alias "uniform" for "random_uniform" and "normal" for "random_normal"

**Comment by asmushetzel** [comment]
Hello Eric, 
after some tries I figured out the root cause of the compile problems:
These changes here need my recent change to mshadow/random.h. This change was integrated into mshadow with PR 228. But it seems that MxNet itself does not use this latest mshadow but instead a fixed older version. So I can't get this stuff here integrated as it will always fail compile.
Do you know a way how we can get this fixed?

**Comment by piiswrong** [comment]
you need to update submodule in your pr

**Comment by asmushetzel** [comment]
So I got all cleaned up except a certain test-failure in roi-pooling. Though this is untouched by these changes here, except for using a newer version of mshadow. It seems that this problem is caused by a recent change on 4/4 (PR 227) which sets kMaxGridNum in tensor_gpu-inl.cuh to MaxInt. The current MxNet master version includes a prior version of mshadow so is not affected by this change. 
My code has to use the more recent mshadow. And in roi-pooling (and one other operator), kMaxGridnum is used in this expression:
dim3 dimGrid(kMaxGridNum, (gridSize + kMaxGridNum - 1) / kMaxGridNum);
which will may to integer overflows. Also dimGrid will be passed to CheckLaunchParam and this will anyway fail then as it expects the first dimension to be smaller than 2^16. 
Will try to fix this.
Any other suggestion (roll back the change in mshadow).

**Comment by sxjscience** [comment]
@precedenceguo We need to revise the launching parameter of ROIPooling.

**Comment by asmushetzel** [comment]
From my side, this PR is fine now. Can we merge it?
Revising launching parameters for ROI pooling as indicated above is a separate task as this PR is not modifying anything concerning ROI-pooling anymore.

File: src/operator/tensor/sample_op.h [hunk 2]
```diff
@@ -167,18 +381,9 @@ inline bool SampleOpType(const nnvm::NodeAttrs& attrs,
 }
 
 inline std::vector<ResourceRequest> SampleResource(const NodeAttrs& attrs) {
-  return { ResourceRequest::kRandom, ResourceRequest::kTempSpace };
```
Comment (piiswrong) [inline src/operator/tensor/sample_op.h]: Are you sure removing this is correct?

**Comment by piiswrong** [comment]
ok. LGTM

**Comment by piiswrong** [comment]
Looks like the name uniform and normal are already take. I don't know how this pr passed the tests but it's failing on my machine on import

**Comment by sxjscience** [comment]
Would it be related to cython API? I can `import mxnet` on windows, which uses ctypes.

**Comment by asmushetzel** [comment]
I used "normal" and "uniform" just as an alias for backward compatibility (see sample_op.cc). Will try tomorrow on my machine (mac) again but last time all tests did pass.

**Comment by asmushetzel** [comment]
I did a complete fresh build on ubuntu and osx and both worked fine.

"uniform" and "normal" were already taken before. There is no real change associated with them except that now they are merely aliases.

**Comment by loofahcus** [comment]
Since the operator names are changed from `uniform` and `normal` to `random_uniform` and `random_normal` in `sample_op.cc`, I think we should also change them in `sample_op.cu` to prevent them from before. See #5890

**Comment by asmushetzel** [comment]
Already done with PR 5883

**Comment by loofahcus** [comment]
Oh, I see.

**Pull request merged**